File: src/cpp/ros_nodes/motor_control_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <std_msgs/msg/string.hpp> // For acknowledgments or status

class MotorControlNode : public rclcpp::Node
{
public:
  MotorControlNode() : Node("motor_control_node")
  {
    // Subscriber to motion commands from the motion planning node
    subscriber_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
      "motion_commands", 10, std::bind(&MotorControlNode::motion_command_callback, this, std::placeholders::_1));

    // Publisher for sending acknowledgments or status updates
    publisher_ = this->create_publisher<std_msgs::msg::String>("motor_control_status", 10);

    RCLCPP_INFO(this->get_logger(), "Motor Control Node started. Subscribed to /motion_commands, publishing on /motor_control_status");
  }

private:
  void motion_command_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
  {
    RCLCPP_INFO(this->get_logger(), "Received motion command. Executing low-level motor control...");
    // TODO: Implement low-level motor control logic here
    // This would involve translating the PoseStamped command into joint torques/velocities

    // Publish a status message indicating command received (and potentially execution status)
    auto status_msg = std_msgs::msg::String();
    status_msg.data = "Motion command executed";
    publisher_->publish(status_msg);
  }

  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr subscriber_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<MotorControlNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: src/motor_control_node.py
# ROS 2 Node for Motor Control

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist

class MotorControlNode(Node):
    def __init__(self):
        super().__init__('motor_control_node')
        self.subscription = self.create_subscription(
            Twist, 'cmd_vel', self.motor_command_callback, 10)
        self.get_logger().info('Motor Control Node Initialized')

    def motor_command_callback(self, msg):
        linear_speed = msg.linear.x
        angular_speed = msg.angular.z
        self.get_logger().info(f'Received motor command: Linear={linear_speed}, Angular={angular_speed}')

        # Placeholder for actual motor control commands
        # This would involve sending commands to motor drivers or a hardware interface
        self.control_motors(linear_speed, angular_speed)

    def control_motors(self, linear, angular):
        # Actual motor control logic would go here.
        # This might involve:
        # - Calculating wheel speeds based on linear and angular velocities
        # - Sending commands to motor controllers (e.g., via serial, I2C, or ROS 2 hardware interfaces)
        # - Implementing safety checks and limits
        self.get_logger().info('Executing motor control...')
        pass

def main(args=None):
    rclpy.init(args=args)
    motor_control_node = MotorControlNode()
    rclpy.spin(motor_control_node)
    motor_control_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
